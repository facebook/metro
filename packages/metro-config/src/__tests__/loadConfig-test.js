/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+javascript_foundation
 * @flow
 * @format
 */

'use strict';

jest.mock('cosmiconfig');

const {loadConfig} = require('../loadConfig');
const getDefaultConfig = require('../defaults');
const cosmiconfig = require('cosmiconfig');

describe('loadConfig', () => {
  beforeEach(() => {
    cosmiconfig.reset();
  });

  it('can load config objects', async () => {
    const config = {
      metro: true,
      reporter: null,
      maxWorkers: 2,
      cacheStores: [],
      transformerPath: '',
    };

    cosmiconfig.setResolvedConfig(config);

    const result = await loadConfig({});

    expect(result).toMatchSnapshot();
    expect(result.cacheStores).toEqual([]);
  });

  it('can load config from function', async () => {
    const config = defaultConfig => ({
      ...defaultConfig,
      cacheStores: [],
      reporter: null,
      maxWorkers: 2,
      resolver: 'test',
      transformerPath: '',
    });

    cosmiconfig.setResolvedConfig(config);

    const result = await loadConfig({});

    expect(result.resolver).toEqual('test');
  });

  it('can load the config with a path', async () => {
    const config = defaultConfig => ({
      ...defaultConfig,
      metro: true,
      reporter: null,
      maxWorkers: 2,
      cacheStores: [],
      transformerPath: '',
    });

    cosmiconfig.setResolvedConfig(config);

    const result = await loadConfig({config: '/metro.config.js'});

    expect(result).toMatchSnapshot();
    expect(cosmiconfig.hasLoadBeenCalled()).toBeTruthy();
  });

  it('can load the config with no config present', async () => {
    cosmiconfig.setReturnNull(true);

    const result = await loadConfig();
    result.reporter = {update: () => {}};

    const defaultConfig = await getDefaultConfig(process.cwd());
    defaultConfig.watchFolders = [
      defaultConfig.projectRoot,
      ...defaultConfig.watchFolders,
    ];
    defaultConfig.reporter = {update: () => {}};

    expect(JSON.stringify(result)).toEqual(JSON.stringify(defaultConfig));
  });
});
