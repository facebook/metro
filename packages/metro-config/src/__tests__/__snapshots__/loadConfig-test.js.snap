// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfig can load config objects 1`] = `
Object {
  "cacheStores": Array [],
  "cacheVersion": "1.0",
  "maxWorkers": 2,
  "metro": true,
  "projectRoot": "/",
  "reporter": null,
  "resetCache": false,
  "resolver": Object {
    "assetExts": Array [
      "bmp",
      "gif",
      "jpg",
      "jpeg",
      "png",
      "psd",
      "svg",
      "webp",
      "m4v",
      "mov",
      "mp4",
      "mpeg",
      "mpg",
      "webm",
      "aac",
      "aiff",
      "caf",
      "m4a",
      "mp3",
      "wav",
      "html",
      "pdf",
      "otf",
      "ttf",
    ],
    "assetTransforms": false,
    "blacklistRE": /\\(node_modules\\[\\\\/\\\\\\\\\\]react\\[\\\\/\\\\\\\\\\]dist\\[\\\\/\\\\\\\\\\]\\.\\*\\|website\\\\/node_modules\\\\/\\.\\*\\|heapCapture\\\\/bundle\\\\\\.js\\|\\.\\*\\\\/__tests__\\\\/\\.\\*\\)\\$/,
    "extraNodeModules": Object {},
    "platforms": Array [
      "ios",
      "android",
      "windows",
      "web",
    ],
    "providesModuleNodeModules": Array [
      "react-native",
      "react-native-windows",
    ],
    "resolveRequest": null,
    "resolverMainFields": Array [
      "browser",
      "main",
    ],
    "sourceExts": Array [
      "js",
      "json",
      "ts",
      "tsx",
    ],
    "useWatchman": true,
  },
  "serializer": Object {
    "createModuleIdFactory": [Function],
    "experimentalSerializerHook": [Function],
    "getModulesRunBeforeMainModule": [Function],
    "getPolyfills": [Function],
    "getRunModuleStatement": [Function],
    "polyfillModuleNames": Array [],
    "postProcessBundleSourcemap": [Function],
    "processModuleFilter": [Function],
  },
  "server": Object {
    "enhanceMiddleware": [Function],
    "port": 8080,
    "useGlobalHotkey": true,
  },
  "transformer": Object {
    "assetPlugins": Array [],
    "assetRegistryPath": "missing-asset-registry-path",
    "asyncRequireModulePath": "metro/src/lib/bundle-modules/asyncRequire",
    "babelTransformerPath": "metro/src/defaultTransformer",
    "dynamicDepsInPackages": "throwAtRuntime",
    "enableBabelRCLookup": true,
    "getTransformOptions": [Function],
    "minifierPath": "metro-minify-uglify",
    "optimizationSizeLimit": 153600,
    "postMinifyProcess": [Function],
    "transformVariants": Object {
      "default": Object {},
    },
    "workerPath": null,
  },
  "transformerPath": "",
  "watch": false,
  "watchFolders": Array [
    "/",
  ],
}
`;

exports[`loadConfig can load the config with a path 1`] = `
Object {
  "cacheStores": Array [],
  "cacheVersion": "1.0",
  "maxWorkers": 2,
  "metro": true,
  "projectRoot": "/",
  "reporter": null,
  "resetCache": false,
  "resolver": Object {
    "assetExts": Array [
      "bmp",
      "gif",
      "jpg",
      "jpeg",
      "png",
      "psd",
      "svg",
      "webp",
      "m4v",
      "mov",
      "mp4",
      "mpeg",
      "mpg",
      "webm",
      "aac",
      "aiff",
      "caf",
      "m4a",
      "mp3",
      "wav",
      "html",
      "pdf",
      "otf",
      "ttf",
    ],
    "assetTransforms": false,
    "blacklistRE": /\\(node_modules\\[\\\\/\\\\\\\\\\]react\\[\\\\/\\\\\\\\\\]dist\\[\\\\/\\\\\\\\\\]\\.\\*\\|website\\\\/node_modules\\\\/\\.\\*\\|heapCapture\\\\/bundle\\\\\\.js\\|\\.\\*\\\\/__tests__\\\\/\\.\\*\\)\\$/,
    "extraNodeModules": Object {},
    "platforms": Array [
      "ios",
      "android",
      "windows",
      "web",
    ],
    "providesModuleNodeModules": Array [
      "react-native",
      "react-native-windows",
    ],
    "resolveRequest": null,
    "resolverMainFields": Array [
      "browser",
      "main",
    ],
    "sourceExts": Array [
      "js",
      "json",
      "ts",
      "tsx",
    ],
    "useWatchman": true,
  },
  "serializer": Object {
    "createModuleIdFactory": [Function],
    "experimentalSerializerHook": [Function],
    "getModulesRunBeforeMainModule": [Function],
    "getPolyfills": [Function],
    "getRunModuleStatement": [Function],
    "polyfillModuleNames": Array [],
    "postProcessBundleSourcemap": [Function],
    "processModuleFilter": [Function],
  },
  "server": Object {
    "enhanceMiddleware": [Function],
    "port": 8080,
    "useGlobalHotkey": true,
  },
  "transformer": Object {
    "assetPlugins": Array [],
    "assetRegistryPath": "missing-asset-registry-path",
    "asyncRequireModulePath": "metro/src/lib/bundle-modules/asyncRequire",
    "babelTransformerPath": "metro/src/defaultTransformer",
    "dynamicDepsInPackages": "throwAtRuntime",
    "enableBabelRCLookup": true,
    "getTransformOptions": [Function],
    "minifierPath": "metro-minify-uglify",
    "optimizationSizeLimit": 153600,
    "postMinifyProcess": [Function],
    "transformVariants": Object {
      "default": Object {},
    },
    "workerPath": null,
  },
  "transformerPath": "",
  "watch": false,
  "watchFolders": Array [
    "/",
  ],
}
`;
